web/tests/filter.test.js

```diff
@@ -2307,3 +2307,53 @@ run_test("adjusted_terms_if_moved", () => {
     result = Filter.adjusted_terms_if_moved(raw_terms, message);
     assert.deepStrictEqual(result, expected);
 });
+
+run_test("can_newly_match_moved_messages", () => {
+    const test_cases = [
+        {
+            name: "matches new stream",
+            new_stream: "new_stream",
+            new_topic: "any_topic",
+            setup(filter) {
+                filter.has_operand = (type, name) => type === "channel" && name === "new_stream";
+            },
+            expected: true,
+        },
+        {
+            name: "matches new topic",
+            new_stream: "any_stream",
+            new_topic: "new_topic",
+            setup(filter) {
+                filter.has_operand = (type, name) => type === "topic" && name === "new_topic";
+            },
+            expected: true,
+        },
+        {
+            name: "matches other conditions",
+            new_stream: "any_stream",
+            new_topic: "any_topic",
+            setup(filter) {
+                filter.has_operand = () => false;
+                filter.sorted_term_types = () => ["is-followed", "channels-public"];
+            },
+            expected: true,
+        },
+        {
+            name: "matches none",
+            new_stream: "any_stream",
+            new_topic: "any_topic",
+            setup(filter) {
+                filter.has_operand = () => false;
+                filter.sorted_term_types = () => ["some_other_term"];
+            },
+            expected: false,
+        },
+    ];
+
+    for (const {name, new_stream, new_topic, setup, expected} of test_cases) {
+        const filter = new Filter([]); // Assuming Filter is the class you're testing
+        setup(filter);
+        const result = filter.can_newly_match_moved_messages(new_stream, new_topic);
+        assert.strictEqual(result, expected, `${name}: ${result} !== ${expected}`);
+    }
+});
```

Comment:
I don't think this is a good way to test this function; we should use actual `Filter` objects, not mock `has_operand`, to do a proper integration test. I guess most of what you've written here might be fine. But fundamentally, if we're doing `new Filter` calls with actual parameters, that's an end-to-end test that will work if we refactor the function slightly, and this logic won't.

(And I think we do want to refactor the function slightly, to do case-insensitive comparisons on the channel/topic).

Reply:
Makes sense, updated!